unit_scaling/_modules.py

```diff
@@ -241,54 +443,111 @@ class TransformerLayer(nn.Module):
         heads (int): the number of attention heads.
-        dropout_p (float, optional): the probability of the post-softmax dropout.
-        act_fn (nn.Module): the activation function module. Defaults to `GELU()`.
-        mhsa_tau (float, optional): the weighting of the multi-head-self-attention
-            branch relative to the skip connection. Defaults to 0.01.
-        mlp_tau (float, optional): the weighting of the MLP
-            branch relative to the skip connection. Defaults to 0.5.
+        mhsa_tau (float): the weighting of the multi-head-self-attention
+            branch relative to the skip connection.
+        mlp_tau (float): the weighting of the MLP
+            branch relative to the skip connection.
+        is_causal (bool): causal masking (for non-padded sequences).
+        dropout_p (float, optional): the probability of residual and post-softmax
+            dropout.
+        act_fn (nn.Module, optional): the activation function module.
+            Defaults to :code:`GELU()`.
         {0}
     """
 
     def __init__(
         self,
         hidden_size: int,
         heads: int,
-        dropout_p: float,
+        mhsa_tau: float,
+        mlp_tau: float,
+        is_causal: bool,
+        dropout_p: float = 0.0,
         act_fn: nn.Module = GELU(),
-        mhsa_tau: float = 0.01,
-        mlp_tau: float = 0.5,
-        constraint: Optional[str] = "gmean",
+        constraint: Optional[str] = "to_output_scale",
     ) -> None:
         super().__init__()
         self.dropout_p = dropout_p
         self.mhsa_tau = mhsa_tau
         self.mlp_tau = mlp_tau
-        self.mhsa_layer_norm = LayerNorm(hidden_size)
-        self.mhsa = MHSA(hidden_size, heads, dropout_p, constraint=constraint)
-        self.mlp_layer_norm = LayerNorm(hidden_size)
+        self.mhsa_norm = RMSNorm(hidden_size)
+        self.mhsa = MHSA(
+            hidden_size,
+            heads,
+            is_causal=is_causal,
+            dropout_p=dropout_p,
+            constraint=constraint,
+        )
+        self.mlp_norm = RMSNorm(hidden_size)
         self.mlp = MLP(hidden_size, act_fn, constraint=constraint)
 
     def forward(self, input: Tensor) -> Tensor:
         input, skip = U.residual_split(input, tau=self.mhsa_tau)
-        input = self.mhsa_layer_norm(input)
+        input = self.mhsa_norm(input)
         input = self.mhsa(input)
         input = U.dropout(input, self.dropout_p, self.training)
         input = U.residual_add(input, skip, tau=self.mhsa_tau)
 
         input, skip = U.residual_split(input, tau=self.mlp_tau)
-        input = self.mlp_layer_norm(input)
+        input = self.mlp_norm(input)
         input = self.mlp(input)
         input = U.dropout(input, self.dropout_p, self.training)
         return U.residual_add(input, skip, tau=self.mlp_tau)
 
 
+@inherit_docstring(
+    "A :class:`torch.nn.Sequential` that automatically configures the depth"
+    " for sake of scaling."
+)
+class Trunk(nn.Sequential):
```

Comment:
using `nn.Sequential` is fine when only passing activations between blocks in the trunk, but typically we pass a lot more (`masks`, `embeddings`). 

Worth considering using a more general `ModuleList` implementation, perhaps with some requirement that elements in list are unit scaled blocks.

Reply:
Agree that this is needed; I've added it alongside, since I think both are valid. Not entirely happy with the name `DepthModuleList`. I wanted to convey that not every `ModuleList` should use this impl, only ones that have a depth-hyperparameter-scaling behaviour, but it seems a bit awkward.

> Perhaps with some requirement that elements in list are unit scaled blocks.

This should already be enforced by the `has_parameter_data` check, I think.